group_project_v2/mixins.py

```diff
@@ -266,18 +266,30 @@ class DashboardRootXBlockMixin(ProjectAPIXBlockMixin):
     TARGET_STUDENTS = 'target_students'
     TARGET_WORKGROUPS = 'target_workgroups'
 
-    def _append_context_parameters_if_not_present(self, context):
+    FILTERED_STUDENTS = "filtered_students"
+
+    def _add_students_and_workgroups_to_context(self, context, filter_by_client_id=None):
         """
-        Appends target students and target workgroups parameters, if not already present in context
-        :param dict context: XBlock view context
-        :rtype: None
+        :param dict context:
+        :param int filter_by_client_id: If none we won't do filtering
+        :return:
         """
-        workgroups, users = self.get_workgroups_and_students()
-        if self.TARGET_STUDENTS not in context:
-            context[self.TARGET_STUDENTS] = users
+        workgroups, students = self.get_workgroups_and_students()
+
+        context[self.TARGET_STUDENTS] = list(students)
+        context[self.TARGET_WORKGROUPS] = list(workgroups)
+        context[self.FILTERED_STUDENTS] = filtered_students = set()
+
+        if filter_by_client_id is None:
+            # Filtering not needed
+            return
+
+        organisation = self.project_api.get_organisation_by_id(filter_by_client_id)
 
-        if self.TARGET_WORKGROUPS not in context:
-            context[self.TARGET_WORKGROUPS] = workgroups
+        for workgroup in workgroups:
+            for user in workgroup.users:
+                if user.id not in organisation.user_ids:
+                    filtered_students.add(user.id)
```

Comment:
@jbzdak nit: this is a bit tricky - on the first read it is easy to think that `filtered_students` are not used, but than, reading it again you'll notice that it is assigned to context variable, and here you're modifying the mutable struct, so when you leave the function the context var will carry all the modifications done here...

But it would be much more intuitive if context var was assigned after the set is built.